tests/unit/test_data.py

```diff
@@ -160,6 +173,53 @@ def test_normalize(self, data_type, weights):
             normalized_with_zeros.weights, jnp.nan_to_num(expected_weights)
         )
 
+    def test_jaxtyping_compatibility(self, data_type):
+        """
+        Test `Data` compatibility with jaxtyping annotations.
+
+        Checks the following cases:
+            - Correct shape and data type,
+            - Correct shape and incorrect data type,
+            - Incorrect shape and correct data type.
+        """
+        float_data_type = eqx.tree_at(
+            lambda x: x.args,
+            data_type,
+            replace=jtu.tree_map(lambda y: jnp.astype(y, jnp.float32), data_type.args),
+        )
+        int_data_type = eqx.tree_at(
+            lambda x: x.args,
+            data_type,
+            replace=jtu.tree_map(lambda y: jnp.astype(y, jnp.int32), data_type.args),
+        )
```

Comment:
We could instead parametrise the dtype for this test here, to avoid code duplication